### ISSUE TITLE
Deploy via GitHub Actions

### ISSUE THREAD

**Issue by Jandev** [Jandev/minifier#4]
The services are now being deployed using Azure DevOps.

Create a build- and release using GitHub Actions

**Comment by AlexNDRmac** [comment]
Hi, I can help with build workflow using GitHub Actions. Will work on that ☝️

**Comment by Jandev** [comment]
Sure thing!

I realized now there aren't any ARM templates (or similar) in this repository. So these also have to be created. You can use any technology you like or are most familiar with, for example, Pulumi, Farmer, Terraform, ARM templates, Azure CLI, etc.

These are all the services deployed right now.
![image](https://user-images.githubusercontent.com/462356/95113586-f84d8d00-0742-11eb-9d07-3b8808623769.png)
I've struck through the services which aren't used (anymore).

All configuration keys can be found in this list:
![image](https://user-images.githubusercontent.com/462356/95113727-2df27600-0743-11eb-939f-0aa3b2b104e3.png)

The values look like this.
```json
[
    {
        "name": "APPINSIGHTS_INSTRUMENTATIONKEY",
        "value": "guid",
        "slotSetting": false
    },
    {
        "name": "AzureWebJobsDashboard",
        "value": "connectionStringToStorage",
        "slotSetting": false
    },
    {
        "name": "AzureWebJobsSecretStorageType",
        "value": "Blob",
        "slotSetting": false
    },
    {
        "name": "AzureWebJobsStorage",
        "value": "connectionStringToStorage",
        "slotSetting": false
    },
    {
        "name": "ClientId",
        "value": "guid",
        "slotSetting": false
    },
    {
        "name": "ClientKey",
        "value": "secret",
        "slotSetting": false
    },
    {
        "name": "CosmosAuthenticationKey",
        "value": "accountkey",
        "slotSetting": false
    },
    {
        "name": "CosmosConnectionString",
        "value": "connectionstringtoCosmos",
        "slotSetting": false
    },
    {
        // This one should be removed as it can be configured via Infrastructure as Code, no need for Key Vault.
        "name": "CosmosConnectionStringSecret",
        "value": "referenceToKeyVaultSecret",
        "slotSetting": false
    },
    {
        // This one should be removed as it can be configured via Infrastructure as Code, no need for Key Vault.
        "name": "CosmosEndpoint",
        "value": "referenceToKeyVaultSecret",
        "slotSetting": false
    },
    {
        "name": "CreateNewUrlsServiceBusConnection",
        "value": "connectionstringtoServiceBus",
        "slotSetting": false
    },
    {
        "name": "FUNCTION_APP_EDIT_MODE",
        "value": "readonly",
        "slotSetting": false
    },
    {
        "name": "FUNCTIONS_EXTENSION_VERSION",
        "value": "~1",
        "slotSetting": false
    },
    {
        "name": "GetNewUrlsServiceBusConnection",
        "value": "connectionStringToServiceBusWithManagePermissions",
        "slotSetting": false
    },
    {
        "name": "letsencrypt:ClientId",
        "value": "guid",
        "slotSetting": false
    },
    {
        "name": "letsencrypt:ClientSecret",
        "value": "secretValue",
        "slotSetting": false
    },
    {
        "name": "letsencrypt:ResourceGroupName",
        "value": "rg-name",
        "slotSetting": false
    },
    {
        "name": "letsencrypt:ServicePlanResourceGroupName",
        "value": "serviceplan-name",
        "slotSetting": false
    },
    {
        "name": "letsencrypt:SiteSlot",
        "value": "",
        "slotSetting": false
    },
    {
        "name": "letsencrypt:SubscriptionId",
        "value": "guid",
        "slotSetting": false
    },
    {
        "name": "letsencrypt:Tenant",
        "value": "myTenant.onmicrosoft.com",
        "slotSetting": false
    },
    {
        "name": "letsencrypt:UseIPBasedSSL",
        "value": "false",
        "slotSetting": false
    },
    {
        "name": "ROUTING_EXTENSION_VERSION",
        "value": "~0.3",
        "slotSetting": false
    },
    {
        "name": "WEBSITE_CONTENTAZUREFILECONNECTIONSTRING",
        "value": "connectionStringToStorage",
        "slotSetting": false
    },
    {
        "name": "WEBSITE_CONTENTSHARE",
        "value": "contentShareValue",
        "slotSetting": false
    },
    {
        "name": "WEBSITE_NODE_DEFAULT_VERSION",
        "value": "6.5.0",
        "slotSetting": false
    }
]
```
Most 'secret' values can be implemented via the infrastructure as code tool. There isn't a specific need for Key Vault (anymore). Probably need some code refactoring to remove the usage of it. 

The most important stuff to build & deploy is:
* The Function App binaries (obviously)
* The Service Bus Namespace + Topic `new-urls`
* Service Plan
* Function App
* Application Insights
* Cosmos DB
* Storage Account (necessary for the Function App)

A little extra, which users of this tool probably want:
As you can probably extract from the configuration values, the LetsEncrypt extension is used over here.
![image](https://user-images.githubusercontent.com/462356/95114523-5e86df80-0744-11eb-9d84-1d4ba78cd421.png)

**Comment by AlexNDRmac** [comment]
@Jandev , let's clarify ... I can make an action that will be triggered when you push the tag, then action will build application in release mode, ... next - the build can be deployed to Azure. But I will need to try the process of deploying the release version of application on my Azure account, after that I can transfer all the steps to the github action. Right?

**Comment by Jandev** [comment]
Sounds about right, yes.

1. When changes are merged to `master`, a release build should be triggered
2. This will also trigger a deployment of the resources to Azure
3. The release build is then deployed to the Function App

If I'm not mistaken, there's a (new) GitHub Action available nowadays which supports ARM Template deployments.

The best way to test the deployment is indeed to deploy it to a (test) subscription.
You could also split this up into 2 separate steps/PR's.
The first one would be to add the additional deployment steps when something is merged to `master`.
The second one would be to create the full template/script/... which is deploying all Azure resources & deploy the application.

**Comment by AlexNDRmac** [comment]
@Jandev could you please add to secrets next params:
```
secrets.AZURE_CREDENTIALS
secrets.AZURE_SUBSCRIPTION_ID
secrets.AZURE_RESOURCE_GROUP
```
It'll be required for deploy via ARM to Azure. I'll open new PR with deploy action today. Thanks

**Comment by Jandev** [comment]
> @Jandev could you please add to secrets next params:
> 
> ```
> secrets.AZURE_CREDENTIALS
> secrets.AZURE_SUBSCRIPTION_ID
> secrets.AZURE_RESOURCE_GROUP
> ```
> 
> It'll be required for deploy via ARM to Azure. I'll open new PR with deploy action today. Thanks

Sure thing, all three have been created.
I'm assuming, based on the naming convention, you're using the [method mentioned in the docs for deployment](https://docs.microsoft.com/en-us/azure/app-service/deploy-github-actions?tabs=userlevel&WT.mc_id=AZ-MVP-5003246#generate-deployment-credentials), right?

**Comment by Jandev** [comment]
Fixed with PR #13
